internal/onelogin/did_test.go

```diff
@@ -0,0 +1,369 @@
+package onelogin
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"encoding/base64"
+	"io"
+	"math/big"
+	"net/http"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+)
+
+var (
+	validKID         = "did:web:identity.integration.account.gov.uk#c9f8da1c87525bb41653583c2d05274e85805ab7d0abc58376c7128129daa936"
+	validDIDDocument = `{
+	"@context" : [ "https://www.w3.org/ns/did/v1", "https://w3id.org/security/jwk/v1" ],
+	"id" : "did:web:identity.integration.account.gov.uk",
+	"assertionMethod" : [ {
+		"type" : "JsonWebKey",
+		"id" : "did:web:identity.integration.account.gov.uk#c9f8da1c87525bb41653583c2d05274e85805ab7d0abc58376c7128129daa936",
+		"controller" : "did:web:identity.integration.account.gov.uk",
+		"publicKeyJwk" : {
+			"kty" : "EC",
+			"crv" : "P-256",
+			"x" : "NPGA7cyIKtH1nz2CJIH14s9_CtC93NwdCQcEi-ADvxg",
+			"y" : "2cTdmHAmZjighly34lXcxEw50cbKFV7FTOdZKhOG7ps",
+			"alg" : "ES256"
+		}
+	} ]
+}`
+
+	validXBytes, _ = base64.RawURLEncoding.DecodeString("NPGA7cyIKtH1nz2CJIH14s9_CtC93NwdCQcEi-ADvxg")
+	validYBytes, _ = base64.RawURLEncoding.DecodeString("2cTdmHAmZjighly34lXcxEw50cbKFV7FTOdZKhOG7ps")
+	validPublicKey = &ecdsa.PublicKey{
+		Curve: elliptic.P256(),
+		X:     new(big.Int).SetBytes(validXBytes),
+		Y:     new(big.Int).SetBytes(validYBytes),
+	}
+)
+
+func TestGetDIDWhenRefreshAfterCacheControlDuration(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	doer := newMockDoer(t)
+	doer.EXPECT().
+		Do(mock.Anything).
+		Return(&http.Response{
+			StatusCode: http.StatusOK,
+			Header: http.Header{
+				"Cache-Control": {"max-age=1"},
+			},
+			Body: io.NopCloser(strings.NewReader(validDIDDocument)),
+		}, nil).
+		Once()
+	doer.EXPECT().
+		Do(mock.Anything).
+		Return(&http.Response{
+			StatusCode: http.StatusOK,
+			Body:       io.NopCloser(strings.NewReader(validDIDDocument)),
+		}, nil).
+		Run(func(req *http.Request) {
```

Comment:
Ah ha - I think this is the step I was missing...so by calling `Run()` you can sequence some actions to happen after this mock does what it needs in the test?